### PR TITLE
Added VEGAS

### PR DESCRIPTION
**Pull request by gomezzz**
Added the first draft of VEGAS.

It is still a little slow as it is not vectorized yet. Created a ticket for that already, but I want to merge this first to take it step by step.

### CODE REVIEW

File: torchquad/integration/vegas_stratification.py [hunk 1]
```diff
@@ -6,22 +6,108 @@
 
 
 class VEGASStratification:
-    """The stratification used for VEGAS Enhanced. Refer to https://arxiv.org/abs/2009.05112."""
+    """The stratification used for VEGAS Enhanced. Refer to https://arxiv.org/abs/2009.05112.
+    Implementation inspired by https://github.com/ycwu1030/CIGAR/ .
+    EQ refers to equation in the paper.
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:11]: Same comments as always ;)

File: torchquad/integration/vegas_stratification.py [hunk 2]
```diff
@@ -6,22 +6,108 @@
...
+    """
 
-    def __init__(self) -> None:
-        pass
+    def __init__(self, dim=1, N_strat=10, beta=0.75):
+        """Initialize the VEGAS stratification
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:15]: "."

File: torchquad/integration/vegas_stratification.py [hunk 3]
```diff
@@ -6,22 +6,108 @@
...
+        self.JF2 = torch.zeros(self.N_cubes)  # jacobian^2 times f
+        self.dh = torch.ones(self.N_cubes) * 1.0 / self.N_cubes  # dampened counts
+        self.strat_counts = torch.zeros(self.N_cubes)  # current index counts
+
+    def accumulate_weight(self, idx, weight):
+        """Accumulate weights for this index + weight
+
+        Args:
+            idx (int): current index
+            weight (float): function value
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:37]: Capit and punct.

File: torchquad/integration/vegas_stratification.py [hunk 4]
```diff
@@ -6,22 +6,108 @@
...
+        """
+        self.JF2[idx] += weight * weight
+        self.JF[idx] += weight
+        self.strat_counts[idx] += 1
 
     def update_DH(self):
-        pass
+        """Update the dampened sample counts"""
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:44]: "."

File: torchquad/integration/vegas_stratification.py [hunk 5]
```diff
@@ -6,22 +6,108 @@
...
+        d_sum = sum(self.dh)
+        self.dh = self.dh / d_sum
 
     def get_NH(self, idx, nevals_exp):
-        pass
+        """Recalculate sample points per hypercube , EQ 44
+
+        Args:
+            idx (int): current index
+            nevals_exp (int): expected number of evaluations
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:69]: No space before a comma. Aaaaand Capi+Punc.

Comment (htoftevaag) [reply torchquad/integration/vegas_stratification.py:69]: Alternatively:
`Recalculate sample points per hypercube (see EQ 44)`

File: torchquad/integration/vegas_stratification.py [hunk 6]
```diff
@@ -6,22 +6,108 @@
...
 
-    def strat_get_indices(self, idx):
-        pass
+        Returns:
+            int: stratified sample counts
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:72]: Capital letter and dot at the end.

File: torchquad/integration/vegas_stratification.py [hunk 7]
```diff
@@ -6,22 +6,108 @@
...
+            return int(nh)
+
+    def _get_indices(self, idx):
+        """Maps point to stratified point
+
+        Args:
+            idx (int): target points index
+
+        Returns:
+            torch.tensor: mapped point
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:87]: You know the drill by now.

File: torchquad/integration/vegas_stratification.py [hunk 8]
```diff
@@ -6,22 +6,108 @@
...
 
     def get_Y(self, idx):
-        pass
+        """Compute randomly sampled points in specified interval
+
+        Args:
+            idx (int): interval index
+
+        Returns:
+            torch.tensor: sampled point
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:105]: Yup, capital letter and dot.

File: torchquad/integration/vegas_stratification.py [hunk 9]
```diff
@@ -6,22 +6,108 @@
...
+        Args:
+            dim (int, optional): Dimensionality. Defaults to 1.
+            N_strat (int, optional): Number of stratification steps per dim. Defaults to 10.
+            beta (float, optional): Beta parameter from VEGAS Enhanced. Defaults to 0.75.
+        """
+        self.dim = dim
+        self.N_strat = N_strat  # stratification steps per dim
+        self.beta = beta  # variable controlling adaptiveness in stratification 0 to 1
+        self.N_cubes = self.N_strat ** self.dim  # total number of subdomains
+        self.V_cubes = (1.0 / self.N_strat) ** self.dim  # volume of hypercubes
```
Comment (htoftevaag) [inline torchquad/integration/vegas_stratification.py:26]: Pablo's comment for himself: In the paper they have 
bilde

**Review by htoftevaag — CHANGES_REQUESTED** [review]
A few quickie comments. In general it looks good

File: torchquad/integration/vegas.py [hunk 10]
```diff
@@ -49,109 +57,200 @@ def integrate(
...
+        x = torch.zeros(self._dim)  # transformed sample points
+        alpha_start = 0.5  # initial alpha value
+        # TODO in the original paper this is adjusted over time
+        self.alpha = alpha_start
 
-        logger.info("Computed integral was " + str(self.results[-1][0]) + ".")
-        return -1
+        # Warmup
+        logger.debug(
+            "|  Iter  |    N_Eval    |     Result     |      Error     |    Acc        | Total Evals"
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply torchquad/integration/vegas.py:172]: This is actually so it align with later following print. Should stay imo.

File: torchquad/integration/vegas.py [hunk 11]
```diff
@@ -49,109 +57,200 @@ def integrate(
...
+        )
+        for warmup_iter in range(5):
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply torchquad/integration/vegas.py:174]: Switched to `warmup_N_it` was by error.

File: torchquad/integration/vegas.py [hunk 12]
```diff
@@ -49,109 +57,200 @@ def integrate(
...
+            sig2 = jf2 / N_samples - pow(jf / N_samples, 2)  # estimated variance
+            self.results[-1] += ih  # store results
+            self.sigma2[-1] += sig2 / N_samples  # store results
+            self.map.update_map()  # adapt the map
+            # TODO fix for integrals close to 0
+            acc = torch.sqrt(
+                self.sigma2[-1] / self.results[-1]
+            )  # compute estimated accuracy,
+            logger.debug(
+                f"|\t{warmup_iter}|         {N_samples}|  {self.results[-1]:5e}  |  {self.sigma2[-1]:5e}  |  {acc:4e}%| {self._nr_of_fevals}"
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply torchquad/integration/vegas.py:199]: image
looks like this, I think it is fine? Honestly I don't want to spend too much time on making debug output pretty :)

File: torchquad/integration/vegas.py [hunk 13]
```diff
@@ -49,109 +57,200 @@ def integrate(
...
-        #     sig2 = jf2 / neval * dV * dV - pow(jf / neval * dV, 2)
-        #     self.results[-1] += ih
-        #     self.sigma2[-1] += sig2 / neval
-
-        # vegas.update_DH()
-        # acc = torch.sqrt(self.sigma2[-1] / self.results[-1])
-
-        # return acc
-        pass
+        """Runs one iteration of VEGAS including stratification (and if enable adaptive map).
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply torchquad/integration/vegas.py:206]: What information are you missing?

File: torchquad/integration/vegas_map.py [hunk 14]
```diff
@@ -1,42 +1,206 @@
...
+                if i > 0:
+                    self.dx_edges[dim][i - 1] = stepsize
+
+        # weights in each intervall
+        self.weights = torch.zeros((self.dim, self.N_intervals))
+        self.smoothed_weights = torch.zeros((self.dim, self.N_intervals))
+        self.summed_weights = torch.zeros(self.dim)  # sum of weights per dim
+        self.delta_weights = torch.zeros(self.dim)  # EQ 11
+        self.avg_weight = torch.zeros(self.dim)  # EQ 14
+        self.std_weight = torch.zeros(self.dim)  # EQ 13
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply torchquad/integration/vegas_map.py:48]: Fixed.

File: torchquad/integration/vegas_map.py [hunk 15]
```diff
@@ -1,42 +1,206 @@
...
+                        self.weights[dim][self.N_intervals - 2]
+                        + 7.0 * self.weights[dim][self.N_intervals - 1]
+                    ) / (8.0 * d_sum)
+                else:
+                    d_tmp = (
+                        self.weights[dim][i - 1]
+                        + 6.0 * self.weights[dim][i]
+                        + self.weights[dim][i + 1]
+                    ) / (8.0 * d_sum)
+                if d_tmp != 0:
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply torchquad/integration/vegas_map.py:151]: I don't see why from the VEGAS Enhanced paper

`(d_tmp - 1.0) / torch.log(d_tmp)` leads to a division of -1 / -INF while the desired value here imo is 0.  Just a numerical consideration to keep in mind for stability.


> Why doesn't it apply to all d_i?

What do you mean this? Or does above answer as well?

File: torchquad/integration/vegas_stratification.py (hunk 9 above)
Comment (gomezzz) [reply torchquad/integration/vegas_stratification.py:26]: Implemented as in the paper now.

File: torchquad/integration/vegas.py [hunk 16]
```diff
@@ -12,33 +12,41 @@
 
 
 class VEGAS(BaseIntegrator):
-    """VEGAS Enhanced in torch. Refer to https://arxiv.org/abs/2009.05112."""
+    """VEGAS Enhanced in torch. Refer to https://arxiv.org/abs/2009.05112.
```
Comment (htoftevaag) [inline torchquad/integration/vegas.py:15]: Add space-dot (" .") after every instance of this link.

File: torchquad/integration/vegas_map.py [hunk 17]
```diff
@@ -7,16 +7,16 @@
...
-    """The map used for VEGAS Enhanced. Refer to https://arxiv.org/abs/2009.05112.
+    """The map used for VEGAS Enhanced. Refer to https://arxiv.org/abs/2009.05112 .
     Implementation is inspired by https://github.com/ycwu1030/CIGAR/ .
-    EQ refers to equation in the paper.
+    EQ <n> refers to equation <n> in the above paper.
     """
 
     def __init__(self, dim, integration_domain, N_intervals=100, alpha=0.5) -> None:
-        """Initializes VEGAS' Enhanced adaptive map
+        """Initializes VEGAS Enhanced' adaptive map
```
Comment (htoftevaag) [inline torchquad/integration/vegas_map.py:16]: It is a bit of a tongue twister, but I still think `VEGAS Enhanced's` is correct?

**Review by htoftevaag — CHANGES_REQUESTED** [review]
Good job.

**Review by GabrieleMeoni — CHANGES_REQUESTED** [review]

**Pull request merged**